### ISSUE TITLE
t.i.fdesc improvements

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2419]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#2419|
|Type|enhancement|
|Created|2007-01-25 02:56:04Z|

1. Add writeToFD, and get rid of duplicate implementations

2. Add tests

3. Fix issues with EINTR

4. Fix ```__all__```

<details><summary>Searchable metadata</summary>

```
trac-id__2419 2419
type__enhancement enhancement
reporter__itamarst itamarst
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1169693764000000 1169693764000000
changetime__1173641918000000 1173641918000000
version__None None
owner__ 
cc__therve cc__exarkun cc__itamarst
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst set owner to @exarkun<br>@itamarst set status to `new`|
|-|-|

Implemented in fdesc-2419, ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

I expect this will also close [#11238](https://github.com/twisted/twisted/issues/11238), but have no Mac to test on.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Which test would demonstrate that [#11238](https://github.com/twisted/twisted/issues/11238) has been fixed?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

test_writeToClosed, or possibly test_writeToInvalid.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

Or, well, technically I'm assuming the untested fact that PTYProcess is using writeToFD, but I didn't have much luck trying to trigger the edge case warner was talking about it in PTYProcess directly last time I worked on this code.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @glyph|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

test_writeAndReadLarge fails consistently for me, under Linux or OS X PPC:


```
[ERROR]: twisted.test.test_fdesc.ReadWriteTestCase.test_writeAndReadLarge

Traceback (most recent call last):
  File "/Users/buildslave/twisted/fdesc-2419/twisted/test/test_fdesc.py", line 75, in test_writeAndReadLarge
    result.append(self.read())
  File "/Users/buildslave/twisted/fdesc-2419/twisted/test/test_fdesc.py", line 48, in read
    return l[0]
exceptions.IndexError: list index out of range
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @itamarst|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst commented|
|-|-|

What version Linux? Which version Python?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve commented|
|-|-|

Every python (2.3/2.4/2.5), under Ubuntu edgy.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve commented|
|-|-|

FYI, readFromFD returns because os.read raises EAGAIN. The following patch resolves the problem:

```
Index: twisted/test/test_fdesc.py
===================================================================
--- twisted/test/test_fdesc.py  (revision 19631)
+++ twisted/test/test_fdesc.py  (working copy)
@@ -44,7 +44,7 @@
     def read(self):
         l = []
         res = fdesc.readFromFD(self.r, l.append)
-        if res is None:
+        if res is None and l:
             return l[0]
         else:
             return res
@@ -72,7 +72,10 @@
         self.failUnless(written > 0)
         result = []
         while len(result) < written:
-            result.append(self.read())
+            r = self.read()
+            if not r:
+                break
+            result.append(r)
         result = "".join(result)
         self.assertEquals(len(result), written)
         self.assertEquals(orig[:written], result)
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve commented|
|-|-|

OK after writing the previous comment I found the problem, the condition of the while is wrong:

```
Index: twisted/test/test_fdesc.py
===================================================================
--- twisted/test/test_fdesc.py  (revision 19631)
+++ twisted/test/test_fdesc.py  (working copy)
@@ -71,7 +71,7 @@
         written = self.write(orig)
         self.failUnless(written > 0)
         result = []
-        while len(result) < written:
+        while len("".join(result)) < written:
             result.append(self.read())
         result = "".join(result)
         self.assertEquals(len(result), written)
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_15"></a>|@itamarst set owner to @therve|
|-|-|

That was exarkun's fault :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve set owner to @itamarst|
|-|-|

Please use the following (new and officially approved :)) format for docstrings: 
```
"""
Doc.
"""
```

The work in the branch is great, but I feel uncomfortable about the _posixserialport change, as it seems basically untested. Thoughts ?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_17"></a>|@itamarst set status to `assigned`|
|-|-|

Well, serial wasn't tested before either, and someone complained about EINTR... let's see what exarkun says.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_18"></a>|@itamarst commented|
|-|-|

exarkun seemed ok with serial change, since it simply getting rid of duplicate code. I'll do the docstrings.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_19"></a>|@itamarst set owner to @therve<br>@itamarst set status to `new`|
|-|-|

Back to you. I'll leave docstrings in process.py to the refactor we'll be doing there.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_20"></a>|@therve set owner to @itamarst|
|-|-|

The branch needs to be merged forward.

Regarding test_writeToFull, I'd be tempted to remove it... If we want to keep it, we may do something like that instead:
```
while result == len(orig):
     result = self.write(orig)
```

Maybe we could add a ticket for adding serial tests. Or for deprecating it.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_21"></a>|@itamarst set owner to @therve|
|-|-|

Ready for review in fdesc-2419-2.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_22"></a>|@therve set owner to @itamarst|
|-|-|

You got a problem when you created the branch: there's a strange 'trunk' directory within, that makes the merge impossible.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_23"></a>|@itamarst set owner to @therve|
|-|-|

How about now?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_24"></a>|@therve set owner to @itamarst|
|-|-|

That's better :). 

I have only one more complain: change the copyright header to 2001-2007 in posixbase, fdesc, and _posixserialport modules.

Then please merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_25"></a>|@itamarst set owner to @therve|
|-|-|

done.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_26"></a>|@therve set owner to @itamarst|
|-|-|

That was OK for me.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun commented|
|-|-|

A lot of conch tests which seem to pass reliably in trunk seem to fail intermittently in this branch.  Maybe the changes are just making an existing intermitent failure more likely.  It'd be good to understand exactly what's going on here before merging, though.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_28"></a>|@therve commented|
|-|-|

There's a problem with a modification in t.i.process: some modules (like t.c.unix) rely on the ProcessExitedAlready import, so we must keep it there.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_29"></a>|@itamarst commented|
|-|-|

No, we should fix those modules. twisted.internet.process is an internal implementation module, it should only be imported by low-level reactor code.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_30"></a>|@itamarst commented|
|-|-|

Should I be fixing this in this branch, or another?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_31"></a>|@therve commented|
|-|-|

I have nothing against fixing it in that branch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_32"></a>|@exarkun commented|
|-|-|

Avoiding gratuitous breakage would be ideal, though.  Can we keep process.ProcessExitedAlready?  It can be deprecated and we can delete it completely later.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_33"></a>|@itamarst set owner to @exarkun|
|-|-|

2.3 slave is now happy, tests pass locally.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_34"></a>|@exarkun set owner to @itamarst|
|-|-|

 * Docstrings missing for test_fdesc.ReadWriteTestCase.{setUp,tearDown,write,read}
 * Bare except in ReadWriteTestCase.tearDown - what exceptions are really expected here?
 * test_writeAndReadLarge, the loop will never terminate if fewer bytes than expected are received
 * Neither the EAGAIN case nor the EINTR case in writeToFD is tested
 * in process.py, ```raise error.ProcessExitedAlready``` should be ```raise error.ProcessExitedAlready()```.
 * in process.py, ProcessWriter.writeSomeData used to raise for any OSError other than EPIPE or EAGAIN.  Now it will never raise, and for any OSError other than EPIPE, EAGAIN, or EINTR, it will return CONNECTION_LOST.  I suppose this is probably correct, but it might make for challenging debugging if an unusual error ever comes back from that write(2).  Not exactly sure what should be done about this.
 * procmon.py has no test coverage.  Strictly speaking, the changes here aren't necessary, since the compatibility name is in place.  It seems pretty unlikely that changing this import is going to introduce any new problems, but it still might be a better idea to omit these from this branch, file a new ticket for the incorrect usage, and let someone who cares enough about runner to fix it all the way deal with it.  Up to you.
 * It is bothersome that the incorrect name used in twisted/conch/unix.py only resulted in intermitent failures.  This suggests that the test coverage for these codepaths is inadequate.  There should probably be another ticket for fixing this. (it's also not clear to me how they ever passed, though, since unix.py should never have been importable).

I ran doc/core/examples/mouse.py and it seems to work still.  Of course, this doesn't test the write path for the serial code, so it doesn't really say anything about the changes. :)

Refactoring the write logic into one function is definitely an improvement.  Thanks for cleaning up docstrings/comments, too.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_36"></a>|@therve removed owner|
|-|-|

I did points 1-5, 7.

The problem with t.c.unix is that we use this import to check if it's available for the platform: the import problem results mostly to skipped tests. I had a quick look at it and nothing obvious came out...

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_38"></a>|@exarkun set owner to @therve<br>@exarkun set status to `new`|
|-|-|

Some trailing whitespace left in fdesc.py, otherwise looks good.  Please merge. :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_39"></a>|@therve set status to `closed`|
|-|-|

(In [19815]) Merge fdesc-2419-2

Author: itamar
Reviewer: exarkun, therve
Fixes [#2419](https://github.com/twisted/twisted/issues/2419)

Improve twisted.internet.fdesc: add writeToFD, add tests, and manage EINTR
errors.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
